R/chunk_req_perform.R

```diff
@@ -0,0 +1,204 @@
+#' Chunk a request
+#'
+#' @inheritParams req_perform
+#' @param chunk_size The size of each chunk.
+#' @param data The data to chunk.
+#' @param apply_chunk A function that applies the chunk to the request. It
+#'   takes two arguments:
+#'
+#'   1. `req`: the original request.
+#'   2. `chunk`: the current data chunk.#'
+#' @return For `req_chunk()` a list of requests. For `chunk_req_perform()` a
+#'   list of parsed responses.
+#' @export
+#'
+#' @examples
+#' req <- request("https://api.restful-api.dev/objects")
+#'
+#' ids <- 1:7
+#' chunk_size <- 3
+#'
+#' apply_chunk <- function(req, chunk) {
+#'   chunk <- rlang::set_names(chunk, "id")
+#'   req_url_query(req, !!!chunk)
+#' }
+#'
+#' parse_resp <- function(resp) {
+#'   parsed <- resp_body_json(resp)
+#'   data.frame(
+#'     id = sapply(parsed, function(x) x$id),
+#'     name = sapply(parsed, function(x) x$name)
+#'   )
+#' }
+#'
+#' # in most cases `chunk_req_perform()` should be sufficient
+#' \dontrun{
+#' responses <- chunk_req_perform(
+#'   req,
+#'   chunk_size = 10,
+#'   data = ids,
+#'   apply_chunk = apply_chunk,
+#'   parse_resp = parse_resp
```

Comment:
I wonder if it's worth adding a `combine_resp` argument too? (which would be `vctrs::vec_rbind()` or similar here). And that makes me wonder if it should be `resp_parse()` and `resp_combine()`.  And then should those become a top-level `req_parse()` that could be used by all the multi-request functions? Then maybe we have `req_perform_XYZ()` for returning lists of responses and `req_{something}_XYZ` for using the parsers?

Reply:
I like the idea. One issue I see is that we basically need three functions:

1. `resp_parse(resp)`: parse the response. The result can be used in `n_pages` and `next_request`.
2. (new) `data(parsed)`: extract the actual data.
3. (new) `combine(data_list)`: combine the data.

An example where 3 functions are required is when the response body looks like `list(nextToken = ..., data = ...)`.

Maybe `resp_parse()` should always return a list where one field is `data`. This way we can avoid adding the 2nd function?

Reply:
Maybe it's worth starting a new issue for this — I think we need to look holistically at all requests that might return multiple responses, i.e. pagination, multi, and chunking. 

Maybe we need an interface more like `req_multi_chunk`, `req_multi_paginate` etc, and then you call something like `req_perform_multi` for all of them. (This function would have to be somewhat different to the current `multi_req_perform()` since it would need to handle pagination where you don't know all the pages up front, and we'd probably want to decouple from performing the requests in parallel).

Reply:
Might be worth looking at python and rust libraries to see how they handle this problem.

Reply:
I looked at a couple of python libraries and they don't have helpers for pagination or chunked requests.
For multiple requests in parallel they simply use the async capabilities (or in some cases threads) that python offers.